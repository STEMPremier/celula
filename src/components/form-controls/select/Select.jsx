import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import { ErrorBox } from '../core';

import SystemIcon from '../../icon/SystemIcon';

import './select.less';

class Select extends Component {
  state = {
    showLabel: false,
    // eslint-disable-next-line react/destructuring-assignment
    selectedValue: this.props.selectedValue,
  };

  handleSelectedValue = event => {
    this.setState({ selectedValue: event.target.value });
  };

  render() {
    const {
      buttonClick,
      buttonIcon,
      buttonShow,
      className,
      disabled,
      errorMsg,
      formId,
      handleChange,
      label,
      name,
      options,
    } = this.props;

    const { showLabel, selectedValue } = this.state;

    const id = `${name}`;

    const classes = cx(
      'ce-select',
      {
        'ce-select--disabled': disabled,
        'ce-select--error': errorMsg,
        'ce-select--show-icon': buttonShow,
      },
      className,
    );

    return (
      <div className={classes}>
        {showLabel && <label htmlFor={id}>{label}</label>}
        <div className="ce-select__container">
          <select
            name={name}
            label={label}
            form={formId}
            id={id}
            disabled={disabled}
            onClick={handleChange}
            value={selectedValue}
            onChange={this.handleSelectedValue}
          >
            <option key={label}>{label}</option>
            {options.map(item => (
              <option key={item.value} value={item.value}>
                {item.name}
              </option>
            ))}
          </select>
          {buttonShow && (
            <button
              className="ce-select__button"
              onClick={buttonClick}
              type="button"
            >
              <SystemIcon
                name={buttonIcon}
                className="ce-select__button__icon"
                color="white"
              />
            </button>
          )}
        </div>
        <div className="ce-select__background" />
        {errorMsg && <ErrorBox errorMsg={errorMsg} />}
      </div>
    );
  }
}

Select.propTypes = {
  /**
   * This function is accesible to the user for clickable events on the icon in the gradient to the right.
   */
  buttonClick: PropTypes.func,
  /**
   * Assign an icon to go in the gradient right hand square.  If you use the buttonShow prop as true, this icon will default to "navigate".  See the <SystemIcon /> for more options on what to enter into this icon prop.
   */
  buttonIcon: PropTypes.string,
  /**
   * The select will appear with the gradient box with icon to the right.
   */
  buttonShow: PropTypes.bool,
  /**
   * A class name added to the select.
   */
  className: PropTypes.string,
  /**
   * Make the selected disabled
   */
  disabled: PropTypes.bool,
  /**
   * The error message in the error box.
   */
  errorMsg: PropTypes.string,
  /**
   * The form the select belongs to.
   */
  formId: PropTypes.string,
  /**
   * A function that is called when changing the `<Select />`.
   */
  handleChange: PropTypes.func.isRequired,
  /**
   * The label is required for accessibilty even if you assign a default value.
   */
  label: PropTypes.string.isRequired,
  /**
   * The name for the select is required for accessibilty purposes of attaching a unique id.
   */
  name: PropTypes.string.isRequired,
  /**
   * The options is the array of objects containing the name and value of each select row.
   */
  options: PropTypes.arrayOf(
    PropTypes.oneOfType([PropTypes.object, PropTypes.string, PropTypes.bool]),
  ).isRequired,
  /**
   * The value is the optional preselected default option to appear in the select.
   */
  selectedValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

Select.defaultProps = {
  buttonClick: () => {},
  buttonIcon: 'navigate',
  buttonShow: false,
  className: '',
  disabled: false,
  errorMsg: '',
  formId: '',
  selectedValue: '',
};

export default Select;
