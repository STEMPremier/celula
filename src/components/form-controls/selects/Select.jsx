/* eslint-disable jsx-a11y/no-noninteractive-tabindex */
import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import SystemIcon from '../../icon/SystemIcon';

import './select.less';

class Select extends React.Component {
  state = {
    showLabel: false,
    // eslint-disable-next-line react/destructuring-assignment
    selectedValue: this.props.selectedValue ? this.props.selectedValue : null,
    // selectedOption: null,
    showArrow: false,
  };

  componentDidMount() {
    // eslint-disable-next-line react/destructuring-assignment
    if (this.props.showArrow) {
      this.setState({
        showArrow: true,
      });
    }
  }
  // componentDidMount() {
  //   if (this.props.selectedValue) {
  //     const selectedObject = this.props.options.filter(
  //       option => option.value === this.props.selectedValue,
  //     );
  //     console.log('selectedObject in componentDidMount', selectedObject);
  //     this.setState({
  //       selectedOption: selectedObject,
  //     });
  //   }
  // }

  handleSelectedValue = event => {
    this.setState(
      {
        selectedValue: event.target.value,
        // selectedOption: this.props.options.filter(
        //   option =>
        //     option.value === event.target.value &&
        //     console.log('inside selecteOptoin', event.target.value),
        // ),
      },
      () => console.log('state after event', this.state),
    );
  };

  render() {
    const {
      className,
      disabled,
      error,
      name,
      form,
      // handleChange,
      label,
      options,
      showArrow,
      rightArrowClick,
      selectedValue,
    } = this.props;

    // eslint-disable-next-line no-unused-vars
    const { showLabel, selectedOption } = this.state;

    const defaultObject = options.filter(
      option => option.value === selectedValue,
    );

    const id = `${name}`;

    const classes = cx(
      'ce-select',
      {
        'ce-select--disabled': disabled,
        'ce-select--error': error,
        'ce-select--show-arrow': showArrow,
      },
      className,
    );

    return (
      <div className={classes}>
        {showLabel && <label htmlFor={id}>{label}</label>}
        {/* <div className="ce-select--wrapper"> */}
        <div className="ce-select--box" tabIndex="0">
          <select
            name={name}
            form={form}
            id={id}
            disabled={disabled}
            onChange={this.handleSelectedValue}
            // tabIndex="0"
          >
            {defaultObject.length ? (
              defaultObject.map(item => (
                <option key={item.value} value={item.value} hidden>
                  {item.name}
                </option>
              ))
            ) : (
              <option>{label}</option>
            )}

            {options.map(item => (
              <option key={item.value} value={item.value}>
                {item.name}
              </option>
            ))}
          </select>
          <div className="ce-select--caret">
            <SystemIcon
              name="down"
              color={disabled ? 'grey' : 'black'}
              className="ce-select--down"
            />
          </div>
          {showArrow && (
            <button
              className="ce-select--outside-arrow"
              onClick={rightArrowClick}
              type="submit"
            >
              <SystemIcon
                name="navigate"
                className="ce-select--arrow"
                color="white"
              />
            </button>
          )}
        </div>
        <div className="ce-select--background-state" />
        {error && (
          <div className="ce-select--error-box-wrapper">
            <div className="ce-select--arrow" />
            <div className="ce-select--error-box">
              <div className="ce-select--error-box-text">{error}</div>
            </div>
          </div>
        )}
        <span error={error} />
      </div>
      // </div>
    );
  }
}

Select.propTypes = {
  /**
   * A class name added to the select.
   */
  className: PropTypes.string,
  /**
   * Make the selected disabled
   */
  disabled: PropTypes.bool,
  /**
   * The select will appear with the right arrow
   */
  showArrow: PropTypes.bool,
  /**
   * The label is required for accessibilty eventhough it will not show if you give it a default value instead.
   */
  label: PropTypes.string.isRequired,
  /**
   * The value is the optional preselected default option to appear in the select.
   */
  selectedValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * The form is what the select belongs to.  NOT SURE EXACTLY.  DOUBLE CHECK WITH IAN.
   */
  form: PropTypes.string,
  /**
   * The name for the select is required for accessibilty purposes of attaching a unique id.
   */
  name: PropTypes.string.isRequired,
  /**
   * A function that is called when changing the `<Select />`.
   */
  // handleChange: PropTypes.func.isRequired,
  /**
   * The error message.
   */
  error: PropTypes.string,
  /**
   * The options is the array of objects containing the name, value, and selected status of each select row.
   */
  // eslint-disable-next-line react/forbid-prop-types
  options: PropTypes.array.isRequired,
  // options: PropTypes.arrayof(PropTypes.oneOfType([PropTypes.object]))
  // .isRequired,
  /**
   * This function is accesible to the user for clickable events on the right arrow in the gradient
   */
  rightArrowClick: PropTypes.func,
};

Select.defaultProps = {
  className: '',
  disabled: false,
  form: '',
  error: '',
  selectedValue: '',
  showArrow: false,
  rightArrowClick: () => {},
};

export default Select;
