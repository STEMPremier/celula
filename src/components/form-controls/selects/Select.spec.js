/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Select from './Select';

describe('<Select />', () => {
  describe('render()', () => {
    test('render the Select component', () => {
      const wrapper = shallow(
        <Select
          handleChange={() => {}}
          label="Select Test Label"
          name="Select Test Name"
          options={['testing options 1', 'testing options 2']}
        />,
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
