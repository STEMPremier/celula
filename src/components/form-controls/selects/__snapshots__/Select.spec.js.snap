// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> render() render the Select component 1`] = `
<div
  className="ce-select"
>
  <div
    className="ce-select--box"
  >
    <select
      disabled={false}
      form=""
      id="Select Test Name"
      name="Select Test Name"
      onChange={[Function]}
      onClick={[Function]}
      value=""
    >
      <option
        key="Select Test Label"
      >
        Select Test Label
      </option>
      <option />
      <option />
    </select>
    <div
      className="ce-select--caret"
    >
      <SystemIcon
        className="ce-select--down"
        color="black"
        name="down"
      />
    </div>
  </div>
  <div
    className="ce-select--background-state"
  />
  <span
    error=""
  />
</div>
`;
